### PR TITLE
Eliminate lexer goroutines

### PR DESCRIPTION
**Pull request by tmthrgd**
This is a pretty large pull request, but it's ultimately rather simple. The `zlexer` and `klexer` goroutines are eliminated and replaced with a new type with a simple chan-like API. This eliminates the goroutines and lexer chans, which will reduce memory usage.

The `scan_rr.go` changes aren't very interesting as they're mostly just find-and-replace.

Errors such as #568/#570 should now be impossible from the `zlexer` code, although similar goroutine leaks can occur from the `ParseZone` code, and in fact have in #786, which needs to be fixed later.

Note: This pull request contains an API change. `io.Reader` errors that were previously ignored in `zlexer` and `klexer` are now surfaced in `ParseZone` and `(*DNSKEY).ReadPrivateKey` respectively.

Also note: `testRR` now panics on error, because that tripped me up multiple times while building this.

```
name       old time/op    new time/op    delta
NewRR-12     15.5µs ± 7%     6.3µs ± 1%  -59.24%  (p=0.000 n=10+10)
ReadRR-12    15.2µs ± 0%     6.7µs ± 1%  -56.14%  (p=0.000 n=7+10)

name       old alloc/op   new alloc/op   delta
NewRR-12     5.22kB ± 0%    0.81kB ± 0%  -84.52%  (p=0.000 n=10+10)
ReadRR-12    5.11kB ± 0%    1.82kB ± 0%  -64.44%  (p=0.000 n=9+10)

name       old allocs/op  new allocs/op  delta
NewRR-12       23.0 ± 0%      17.0 ± 0%  -26.09%  (p=0.000 n=10+10)
ReadRR-12      23.0 ± 0%      19.0 ± 0%  -17.39%  (p=0.000 n=10+10)
```

Updates #786
Updates coredns/coredns#2194

### CODE REVIEW

**Comment by tmthrgd** [comment]
@miekg Almost all the memory allocated in NewRR (all but 0.98KB per run) is from [`bufio.NewScanner`](https://github.com/golang/go/blob/389e942745ddd7eef44b71571c463b0dfc3dcac2/src/bufio/bufio.go#L60-L63) and it's default [4KiB buffer](https://github.com/golang/go/blob/389e942745ddd7eef44b71571c463b0dfc3dcac2/src/bufio/bufio.go#L18). That seems excessive to me.

**Comment by miekg** [comment]
J*sus, your're fast. Let me review :)

bufio.NewScanner seems a big hit indeed.

**Comment by tmthrgd** [comment]
I've addressed the `bufio.Reader` allocation by special-casing `io.ByteReader` and reducing the buffer size where it's needed. We read byte-by-byte and any type that implements `io.ByteReader` shouldn't need the extra buffering anyway. That drastically cuts down on the memory use and gives a nice speed-up.

> J*sus, your're fast. Let me review :)

It's a worthwhile Sunday afternoon. :grin:

**Comment by tmthrgd** [comment]
Hold off merging, I screwed up the case of comments inside braces I think. I need to review that part a little closer. It’s not covered by any tests at the moment.

File: dnssec_keyscan.go [hunk 1]
```diff
@@ -219,66 +209,140 @@ func parseKey(r io.Reader, file string) (map[string]string, error) {
...
-	var l lex
-	str := "" // Hold the current read text
-	commt := false
-	key := true
-	x, err := s.tokenText()
-	defer close(c)
-	for err == nil {
-		l.column = s.position.Column
-		l.line = s.position.Line
+func (kl *klexer) Next() (lex, bool) {
```
Comment (miekg) [inline dnssec_keyscan.go]: The comment above it not true anymore.

Comment (tmthrgd) [reply dnssec_keyscan.go]: Good catch, I'll remove the comment from both of them.

**Comment by tmthrgd** [comment]
@miekg What should the behaviour of this code be?
```go
func TestParseBracedComments(t *testing.T) {
	r := strings.NewReader(`name. IN SOA  a6.nstld.com. hostmaster.nic.name. (
		203362132 ; serial
		5m        ; refresh (5 minutes)
		5m        ; retry (5 minutes)
		2w        ; expire (2 weeks)
		300       ; minimum (5 minutes)
	) ; y
	. 3600000  IN  NS ONE.MY-ROOTS.NET. ; x`)
	for r := range ParseZone(r, "", "") {
		t.Log(r.Comment)
	}
}
```

Without the `; y` comment it prints:
```
=== RUN   TestParseBracedComments
--- PASS: TestParseBracedComments (0.00s)
    parse_test.go:352:
    parse_test.go:352:
```

With the `; y` comment it prints:
```
=== RUN   TestParseBracedComments
--- PASS: TestParseBracedComments (0.00s)
    parse_test.go:352: ; serial ; refresh (5 minutes) ; retry (5 minutes) ; expire (2 weeks) ; minimum (5 minutes) ; y
    parse_test.go:352:
```

It loses the `; x` comment because it's at the end of a line and it loses the comment in the braces if there isn't a comment after the braces. Does it matter and is this intentional?

**Comment by miekg** [comment]
The "returning comments" seemed like a must have feature at the time, but I doubt it is useful. There is also no spec that says what it should do, so anything goes.

**Comment by tmthrgd** [comment]
I was hoping I could rip out all the comment handling in a follow-up pull request, because it's somewhat messy and dicey, but unfortunately it's relied upon by [barnybug/cli53](https://github.com/barnybug/cli53/blob/d6f1c68a757d81dd3e8b3ecbb7caf82109232592/bind.go#L15-L50).

**Comment by tmthrgd** [comment]
@miekg Unless there's anything you want me to change, I believe this is now done and ready to merge.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Eliminate lexer gor..." ]
>I was hoping I could rip out all the comment handling in a follow-up pull request, because it's somewhat messy and dicey, but unfortunately it's relied upon by [barnybug/cli53](https://github.com/barnybug/cli53/blob/d6f1c68a757d81dd3e8b3ecbb7caf82109232592/bind.go#L15-L50).

yeah, stuff like this is why I added it, at the time pool.ntp.org wanted such a 
feature.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
coredns test suite also runs OK using this pr.

LGTM

**Comment by tmthrgd** [comment]
@miekg Shall I merge?

**Comment by miekg** [comment]
Go ahead

On 15 Oct 2018 08:06, "Tom Thorogood" <notifications@github.com> wrote:

@miekg <https://github.com/miekg> Shall I merge?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/miekg/dns/pull/792#issuecomment-429731155>, or mute the
thread

**Pull request merged**